### PR TITLE
Add hyphy

### PR DESCRIPTION
**Pull request by stevenweaver**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -113,3 +113,6 @@
 [submodule "tools/wgsim/src"]
 	path = tools/wgsim/src
 	url = https://github.com/lh3/wgsim.git
+[submodule "tools/hyphy/src"]
```
Comment (robertaboukhalil) [inline .gitmodules:116]: Nitpick: Could you move this block up after `htslib` so this file is alphabetically sorted 🙂

File: biowasm.json [hunk 2]
```diff
@@ -356,6 +356,22 @@
 				}
 			]
 		},
+		{
+			"name": "hyphy",
+			"programs": ["hyphy"],
+			"description": "HyPhy",
```
Comment (robertaboukhalil) [inline biowasm.json:362]: Could you add a brief description here, e.g. `Hypothesis testing using phylogenies`. That description will show up on the [package listing page](https://biowasm.com/cdn/v3)

File: biowasm.json [hunk 3]
```diff
@@ -356,6 +356,22 @@
...
+				{
+					"version": "1.0.9",
+					"branch": "v1.0.9"
+				},
+				{
+					"version": "1.0.11",
+					"branch": "v.1.0.11"
+				}
+			]
+		},
```
Comment (robertaboukhalil) [inline biowasm.json:374]: This is so biowasm knows which branch to use to fetch which version of Hyphy
```suggestion
			"versions": [
				{
					"version": "2.5.52",
					"branch": "2.5.52"
				}
			]
		},
```

File: tools/hyphy/examples/v.1.0.11.html [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<script src="https://biowasm.com/cdn/v3/aioli.js"></script>
```
Comment (robertaboukhalil) [inline tools/hyphy/examples/v.1.0.11.html:1]: You can delete those 2 html files and make one called `tools/hyphy/examples/2.5.52.html` with an example HyPhy command.

Then you can modify `web/tests/test_cdn.cy.js` to add a line for hyphy/2.5.52 and which output you expect to see when running that command

File: tools/hyphy/compile.sh [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Comment (robertaboukhalil) [inline tools/hyphy/compile.sh:1]: Hmm, this gives me errors like:

```
/src/biowasm/tools/hyphy/src/src/core/matrix.cpp:3431:50: error: use of undeclared identifier 'vld1q_lane_f64'
                                                 vld1q_lane_f64 (storage.theData + s0, R1, 0), 1);
                                                 ^
...
```

Curious which version of Emscripten you're using?

I'm still on 2.0.25 (https://github.com/biowasm/biowasm/blob/main/CONTRIBUTING.md#development-setup)

File: tools/hyphy/compile.sh [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+
+# Remove extra space in front of Emscripten "-s" variables to avoid errors
+emcmake cmake -DCMAKE_EXE_LINKER_FLAGS="${EM_FLAGS//-s /-s} --preload-file data@/hyphy" .
```
Comment (robertaboukhalil) [inline tools/hyphy/compile.sh:4]: Did you mean to preload folder `tests/data`? I don't see a `data/` folder in the root path
```suggestion
emcmake cmake -DCMAKE_EXE_LINKER_FLAGS="${EM_FLAGS//-s /-s} --preload-file tests/data@/hyphy" .
```

**Review by robertaboukhalil — COMMENTED** [review]
@stevenweaver Thanks so much for submitting this! I added a few comments/questions.

File: tools/hyphy/compile.sh (hunk 5 above)
Comment (stevenweaver) [reply tools/hyphy/compile.sh:1]: I had to make an edit to our CMake configuration in order for this to work. Let me see if I can set a flag.

Comment (stevenweaver) [reply tools/hyphy/compile.sh:1]: Support has been added in HyPhy 2.5.3

File: tools/hyphy/compile.sh (hunk 6 above)
Comment (stevenweaver) [reply tools/hyphy/compile.sh:4]: Remnant of copying tn93 over. I've removed the statement.

File: tools/hyphy/examples/v.1.0.11.html (hunk 4 above)
Comment (stevenweaver) [reply tools/hyphy/examples/v.1.0.11.html]: It's completed, but I'm unsure how to test the page with Aioli. I did make sure that the compiled version of hyphy ran `help` and `version` correctly.

**Comment by stevenweaver** [comment]
Thanks for your feedback. I went ahead and made the requested adjustments. I still do need to see if it will run a complete analysis, but everything compiles and `help` and `version` runs with `node 20`.

**Comment by stevenweaver** [comment]
I was able to get UnitTests to work by using preload-files. 

```
17:29:21 sweaver@Stevens-MacBook-Pro.local src master ? node ../build/hyphy.js hyphy/hbltests/UnitTests/HBLCommands/LUSolve.bf                                                                                                                   7 ↵
[Running COVERAGE TEST 'LUSolve']
[TEST PASSED]
```

Running commands like 

```
node ../build/hyphy.js busted 
```

Does not work. Either advanced path manipulation or code changes within hyphy will need to occur to support it. 

Best,
Steven

File: tools/hyphy/compile.sh [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+EM_FLAGS="-sSTACK_SIZE=2097152 -g4 -sASSERTIONS=1"
```
Comment (robertaboukhalil) [inline tools/hyphy/compile.sh:3]: Once you're done testing, it's worth replacing `-g4` with `-O2` so Emscripten produces smaller .wasm/.js files

**Review by robertaboukhalil — COMMENTED** [review]
Ah yes so because of EM_FLAGS, it's not really meant to be used with Node.js. The best way to test it is by setting up Aioli and modifying these files:
* https://github.com/biowasm/aioli/blob/main/src/examples/test-biowasm.html
* https://github.com/biowasm/aioli/blob/main/src/examples/test-biowasm.js

If you add --preload-file during compilation to preload files in the repo, you can use those files to test running commands. Or you can mount local files from your computer at runtime using Aioli's `mount()` function ([docs here](https://biowasm.com/documentation#local-files))

Hope this helps!

File: tools/hyphy/compile.sh (hunk 7 above)
Comment (robertaboukhalil) [reply tools/hyphy/compile.sh:3]: Also, which version of Emscripten are you using? I'm getting `Attempt to set a non-existent setting: STACK_SIZE`. It's best to use the Emscripten image mentioned in https://github.com/biowasm/biowasm/blob/main/CONTRIBUTING.md

Comment (stevenweaver) [reply tools/hyphy/compile.sh:3]: I'm using version 3.1.44. Unfortunately, I do get the error 

```
10:32:46 sweaver@Stevens-MacBook-Pro.local build hyphy ? node ./hyphy.js 

Stack overflow detected.  You can try increasing -sSTACK_SIZE (currently set to 65536)
/Users/sweaver/Programming/biowasm/tools/hyphy/build/hyphy.js:129
      throw ex;
      ^
```

If it is not set. 

In emscriptem 2, it appears the analogous flag is `EMFLAGS="-sTOTAL_STACK=2097152"`.

**Comment by stevenweaver** [comment]
I'm having some troubles getting my aioli environment working, no matter which way I try. 

1) I tried serving directly from the tool directly: 

`/Users/sweaver/Programming/biowasm/tools/hyphy`. There is an issue with aioli trying to retrieve `hyphy.data` unsuccessfully (due to URL parsing issue, not for lack of existence). 

Screenshot 2023-08-23 at 12 01 52

2) I tried using the aioli example directly, and serving from two separate ports. `hyphy.js` won't load at all, despite being able to navigate to it in my browser and no CORS issue. 

Screenshot 2023-08-23 at 12 02 29


Are there any more tips so that I can move forward? 

Best,
Steven

File: tools/hyphy/compile.sh [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+EM_FLAGS="-sTOTAL_STACK=2097152 -02 -sASSERTIONS=1"
+
+# Remove extra space in front of Emscripten "-s" variables to avoid errors
+emcmake cmake -DCMAKE_EXE_LINKER_FLAGS="${EM_FLAGS//-s /-s} --preload-file res@/hyphy --preload-file tests@/hyphy" .
+emmake make hyphy
+emmake make install
+mv hyphy.{js,wasm} ../build/
```
Comment (robertaboukhalil) [inline tools/hyphy/compile.sh:9]: This is missing .data, I think that's why it's giving you errors. If I make this change, it seems to work:
```suggestion
mv hyphy.{js,wasm,data} ../build/
```

**Review by robertaboukhalil — COMMENTED** [review]
Thanks Steven! I added a few comments, those changes seem to make it possible for me to run `hyphy --version`.

Sidenote: I do see a bunch of stderr outputs. If you want to ignore those, you can set `printInterleaved` to false (see [docs](https://biowasm.com/documentation#aioli-init-advanced)) such that CLI.exe returns an object that splits up stdout/stderr: `{"stdout": "...", "stderr": "..." }`

**Comment by robertaboukhalil** [comment]
Hey @stevenweaver, just following up on this, are you able to run it in the browser and run small analyses? Let me know if I can help with anything!

**Comment by stevenweaver** [comment]
Hi @robertaboukhalil, 

I was able to run hyphy in the browser, but `hyphy` operates by reading in `batch files` in its domain-specific language. I think what needs to be done before going forward is to find a way load and execute these `.bf` files. My bandwidth is a bit low at the moment, so I haven't had a chance to give a go since we've last corresponded. 

Best,
Steven

**Comment by robertaboukhalil** [comment]
Here's sample code for mounting local user files to the virtual filesystem: https://biowasm.com/documentation#local-files. This returns a file path, which you can pass to `hyphy`.

If you want these `.bf` files to be available without users having to select files from their computer, you can [load them from a URL](https://biowasm.com/documentation#remote-files) and have them preloaded in the app. Or if they're small enough you can preload them with `--preload-file` at compilation time.

Hope this helps!

**Comment by robertaboukhalil** [comment]
Hey @stevenweaver! From [this tweet](https://twitter.com/sergeilkp/status/1719142229791379646), sounds like hyphy is compiling! I'd love to get this into biowasm, let me know if I can help.

**Comment by stevenweaver** [comment]
Dear @robertaboukhalil, 

I've released `hyphy 2.5.56`, which includes better support for biowasm. The latest commit shows which CMake flags work and was used with the Observable notebook. If what is in `tools/hyphy` looks good, then it appears we can try to work to see if we can either integrate the Observable notebook or port over applicable JavaScript to the biowasm repo?

Best,
Steven

**Comment by robertaboukhalil** [comment]
That's awesome, thanks Steven, I will take a look this week!

> If what is in tools/hyphy looks good, then it appears we can try to work to see if we can either integrate the Observable notebook or port over applicable JavaScript to the biowasm repo?

Ah sorry, in my previous comment, I just meant bringing `hyphy.wasm` into biowasm (as this PR does), so developers can get the tool from the biowasm.com CDN. No need to bring your observable app into biowasm, it makes sense to keep that separate.

**Comment by robertaboukhalil** [comment]
Hi Steven!

It's looking good so far, just a few questions:

1. The `hyphy.data` file is 16MB; do we always need it preloaded when using hyphy? Or can we not include it for the biowasm release, and users can mount the subset of files they need?

2. The hyphy.wasm output is really large, 12MB (I see the observable tool has the same issue). I was able to reduce is to 2.6MB by not including `-g` in the compilation:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3aae56cd..f5bd3f87 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -631,11 +631,11 @@ set_target_properties(
     hyphy
     PROPERTIES
     if(${OPENMP_FOUND})
-        COMPILE_FLAGS "${DEFAULT_COMPILE_FLAGS} ${DEFAULT_WARNING_FLAGS} ${OpenMP_CXX_FLAGS} -g "
-        LINK_FLAGS "${DEFAULT_LINK_FLAGS} ${OpenMP_CXX_FLAGS} -g "
+        COMPILE_FLAGS "${DEFAULT_COMPILE_FLAGS} ${DEFAULT_WARNING_FLAGS} ${OpenMP_CXX_FLAGS} "
+        LINK_FLAGS "${DEFAULT_LINK_FLAGS} ${OpenMP_CXX_FLAGS} "
     else(${OPENMP_FOUND})
-        COMPILE_FLAGS "${DEFAULT_COMPILE_FLAGS} ${DEFAULT_WARNING_FLAGS} -g "
-        LINK_FLAGS "${DEFAULT_LINK_FLAGS} -g "
+        COMPILE_FLAGS "${DEFAULT_COMPILE_FLAGS} ${DEFAULT_WARNING_FLAGS} "
+        LINK_FLAGS "${DEFAULT_LINK_FLAGS} "
     endif(${OPENMP_FOUND})
 )
```

You could modify hyphy to only add `-g` when the system is not Emscripten, or I can add this as a patch in biowasm if that's easier for you.

**Comment by stevenweaver** [comment]
Thanks @robertaboukhalil, 

I think it would be best to remove `-g` from CMake. It definitely shouldn't be there. I'll do that shortly. As far `hyphy.data`, I'm sure test FASTA files must be finding their way in there if it's that large. I'll take a look shortly. 

Best,
Steven

**Comment by stevenweaver** [comment]
Adjustment to CMake are [here](https://github.com/veg/hyphy/pull/1665) and is part of release `2.5.57`

**Review by robertaboukhalil — APPROVED** [review]

**Comment by robertaboukhalil** [comment]
Thanks @stevenweaver, it looks great! It's now available on the biowasm CDN: https://biowasm.com/cdn/v3/hyphy/2.5.57

> Is there a maximum threshold we should be aiming for or just trying to reduce as much as possible?

There is no threshold, but I think it's good now, the total assets are ~8MB instead of 28MB.

Also I trimmed down the hyphy description so it's more similar in length to the other tools, hope it looks ok.

**Pull request merged**